core/src/main/java/org/apache/accumulo/core/metadata/Reference.java

```diff
@@ -38,4 +38,28 @@ public Reference(TableId tableId, String metadataEntry) {
     this.metadataEntry = metadataEntry;
     this.tabletDir = metadataEntry;
   }
+
+  @Override
+  public int compareTo(Reference that) {
+    if (equals(that)) {
+      return 0;
+    } else {
+      return this.metadataEntry.compareTo(that.metadataEntry);
+    }
+  }
+
+  @Override
+  public boolean equals(Object obj) {
+    if (obj instanceof Reference) {
+      Reference that = (Reference) obj;
+      return this.metadataEntry.equals(that.metadataEntry);
+    } else {
+      return false;
+    }
+  }
```

Comment:
Eclipse generates the following for equals here. My guess is that it might be safer and faster.

```
    if (this == obj)
      return true;
    if (obj == null)
      return false;
    if (getClass() != obj.getClass())
      return false;
    Reference other = (Reference) obj;
    if (metadataEntry == null) {
      if (other.metadataEntry != null)
        return false;
    } else if (!metadataEntry.equals(other.metadataEntry))
      return false;
    return true;
```

Reply:
OK, that is definitely more thorough. I will replace it. 

As far as your comment about using all the fields, theoretically, the other 2 fields _should_ be contained in `metadataEntry`. I am not 100% positive though cause the code is very complicated.

Reply:
Yeah, I agree. It seems to me that the other two fields could be parsed out of the metadata entry, but that's only the case if the entry is *not* relative. I remember many versions ago that the file references were not fully qualified in the metadata table. When we added the ability to support multiple volumes @ericnewton made all of the file references fully qualified. I think it was out of need, but also being specific probably made the code a little easier. I'm surprised that we went backwards to *not* using fully qualified references.

Reply:
So that is for the typical tablet file references, which I created `TabletFile` to represent which look like `hdfs://1.2.3.4/accumulo/tables/2a/t-0003/C0004.rf`. The GC is a different story entirely. The GC uses Strings to represent relative paths (for file and directories) as an optimization. I created the `Reference` class as a replacement for these Strings in the GC.

For example, the GC will create relative paths to pass around like:
https://github.com/apache/accumulo/blob/76aaba0ec4e086ab535a1e88ba6ecb0f62481b2c/server/gc/src/test/java/org/apache/accumulo/gc/GarbageCollectionTest.java#L319-L321

And directories like:
https://github.com/apache/accumulo/blob/76aaba0ec4e086ab535a1e88ba6ecb0f62481b2c/server/gc/src/test/java/org/apache/accumulo/gc/GarbageCollectionTest.java#L374-L379

Reply:
So you are saying that Reference and it's subclasses are *not* used as file references in the metadata table? I got confused by the class locations / packages / names and the other issues that you referenced with prior changes. If that's the case, then can we move these classes out of core and into the GC module. If not, at the very least, can we change their package. It would appear that these classes are used in the metadata table.

Reply:
These changes are building on this change here: https://github.com/apache/accumulo/pull/2688

Reply:
To be clear, my only issue is that it appears that relative things are being used in the metadata table due to class location. If only the GC process is creating relative file reference objects for use at runtime, then you would only know that by tracing through the code to see where things are used / created.

Reply:
So I can move `Reference` to a package in `core`. It can't be in `server.base` or `server.gc` because it is used in `Ample` which is in `org.apache.accumulo.core.metadata.schema`. I was thinking of creating a new package `org.apache.accumulo.core.gc` to put the classes in...

Reply:
> I was thinking of creating a new package org.apache.accumulo.core.gc

Something like that would be great IMO.